### PR TITLE
Extends ports export

### PR DESCRIPTION
**Pull request by nagyandras95**
This PR contains:

- Export missing port actions to UML and C++ (send, connect)
- Export interfaces to c++ (provided, required parts)
- Export port fields with interfaces
- Also contains minor refactoring
- Add configuration to PingPong demo example, which is able to be exported for now, add it to integration compile and runtime tests.

### CODE REVIEW

File: dev/plugins/hu.elte.txtuml.export.cpp/cpp-runtime/InterfaceUtils.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+class EmptyReqInf {
+
+protected:
+	virtual void sendAny(ES::EventRef s) = 0;
+};
+
+class EmptyProvInf {
+protected:
+	virtual void reciveAny(ES::EventRef s) = 0;
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.cpp/cpp-runtime/InterfaceUtils.hpp]: Typo, should be _receive_.

File: dev/plugins/hu.elte.txtuml.export.cpp/cpp-runtime/PortUtils.hpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+{
+public:
+	template <typename RequiredInf1, typename ProvidedInf1>
+	friend class PortImpl;
+
+	template <typename RequiredInf1, typename ProvidedInf1>
+	friend struct AssemblyConnection;
+
+	void setAssemblyConnectedPort (ES::IPortRef<RequiredInf, ProvidedInf> connectedPort_);
+	void setDelgationConnectedPort (ES::PortRef<ProvidedInf, RequiredInf> connectedPort_);
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.cpp/cpp-runtime/PortUtils.hpp]: Typo, should be `Delegation`.

File: dev/plugins/hu.elte.txtuml.export.cpp/cpp-runtime/PortUtils.hpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	PortType getType() const { return type; }
+protected:
+
+	PortType type;
+	ES::ModelObjectRef owner;
+};
+
+struct IConnection
+{
+	virtual void fowardSendedMessageToConnectedPort(ES::EventRef signal) = 0;
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.cpp/cpp-runtime/PortUtils.hpp]: Typo, should be _forwardSent_ (or maybe just _forwarded_?).

File: dev/plugins/hu.elte.txtuml.export.cpp/cpp-runtime/runtime.cpp [hunk 4]
```diff
@@ -127,3 +127,9 @@ void ConfiguredThreadedRT::setConfiguration(ESContainer::FixedArray<ES::SharedPt
 
 }
 
+
+// Constans
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.cpp/cpp-runtime/runtime.cpp]: Typo, should be _constant_.

File: dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/structural/PortExporter.java [hunk 5]
```diff
@@ -1,38 +1,141 @@
...
+		for (Port port : ports) {
+			Pair<String,String> interaces = getPortActualInterfaceTypes(port);
+			usedInterfaces.add(interaces.getFirst());
+			usedInterfaces.add(interaces.getSecond());
+		}
+		
+		return usedInterfaces;
+	}
+	
+	public String crearePortRelatedCodes() {
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/structural/PortExporter.java]: Typo, should be _create_.

File: dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/structural/PortExporter.java [hunk 6]
```diff
@@ -1,38 +1,141 @@
...
+	public String createPortTypeInfoDefinitions() {
+		StringBuilder source = new StringBuilder("");
+		for(Port port : ports) {
+			source.append(PortTemplates.portTypeInfoDef(port.getName(), ownerName));
+		}
+		
 		return source.toString();
 	}
+	
+	private String crteaInterfacePortDefinitionCode(Port port) {
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/structural/PortExporter.java]: Typo, should be _create_.

File: dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/structural/PortExporter.java [hunk 7]
```diff
@@ -1,38 +1,141 @@
...
+		List<String> parameters = new ArrayList<>();
+		if(port.isBehavior()) {
+			parameters.add(PortTemplates.portTypeInfoName(port.getName(), ownerName));
+			parameters.add(GenerationNames.PointerAndMemoryNames.Self);
+		}
+		return ObjectDeclDefTemplates.setAllocatedObjectToObjectVariable(portTypeName, 
+				Optional.of(Arrays.asList(interfaces.getFirst(),interfaces.getSecond())), 
+				port.getName(), Optional.of(parameters), true);
+	}
+	private String createinterfacePortCode(Port port) {
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/structural/PortExporter.java]: Typo, should be _createInterface_.

File: dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/templates/GenerationNames.java [hunk 8]
```diff
@@ -112,23 +119,52 @@
...
 	public static class StateMachineMethodNames {
+
 		public static final String InitFunctionName = "init";
 		public static final String ProcessEventFName = "process_event";
 		public static final String InitializeFunctionName = "initialize";
 		public static final String FinalizeFunctionName = "finalize";
 	}
-	
+
+	public static class TypeDelcreationKeywords {
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/templates/GenerationNames.java]: Typo, should be _Declaration_.

File: dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/templates/GenerationNames.java [hunk 9]
```diff
@@ -112,23 +119,52 @@
...
+		public static final String SendReceptionName = "send";
+		public static final String ReciveReceptionName = "recive";
+		public static final String CommonSendAnySignalName = "sendAny";
+		public static final String CommonReciveAnySignalName = "reciveAny";
+		public static final String IntegratedBaseTemplateName = "IntegratedInf";
+		public static final String EmptyInfName = "EmptyInf";
+
+	}
+
+	public static class InitiliazetFixFunctionNames {
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/templates/GenerationNames.java]: Typo, maybe _Initializer_?

File: dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/templates/GenerationNames.java [hunk 10]
```diff
@@ -112,23 +119,52 @@
...
+		public static final String PrivatePartSign = "private:";
+
+	}
+
+	public static class InterfaceNames {
+
+		public static final String SendReceptionName = "send";
+		public static final String ReciveReceptionName = "recive";
+		public static final String CommonSendAnySignalName = "sendAny";
+		public static final String CommonReciveAnySignalName = "reciveAny";
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/templates/GenerationNames.java]: Typo, should be _Receive_.

File: dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/templates/GenerationNames.java [hunk 11]
```diff
@@ -112,23 +119,52 @@
...
+		public static final String PublicPartSign = "public:";
+		public static final String ProtectedPartSign = "protected:";
+		public static final String PrivatePartSign = "private:";
+
+	}
+
+	public static class InterfaceNames {
+
+		public static final String SendReceptionName = "send";
+		public static final String ReciveReceptionName = "recive";
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/templates/GenerationNames.java]: Typo, should be _Receive_.

File: dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/templates/GenerationNames.java [hunk 12]
```diff
@@ -255,11 +290,17 @@ private static final String simpleProcessEventDefBody() {
 		return "{\n" + "bool handled=false;\n" + "auto range = " + TransitionTableName + ".equal_range(EventState("
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.export.cpp/src/hu/elte/txtuml/export/cpp/templates/GenerationNames.java]: These definitions are still horrendous, please consider Xtend's [template expressions](http://www.eclipse.org/xtend/documentation/203_xtend_expressions.html#templates).

File: dev/plugins/hu.elte.txtuml.utils.eclipse/src/hu/elte/txtuml/utils/jdt/SharedUtils.java [hunk 13]
```diff
@@ -102,11 +103,49 @@ public static Annotation obtainAnnotation(BodyDeclaration declaration, Class<?>
 		return null;
 	}
 	
-	public static ITypeBinding obtainTypeLiteralAnnotation(BodyDeclaration declaration, Class<?> annotationClass) {
-		Expression expr = obtainSingleMemberAnnotationValue(declaration, annotationClass);
-		if (expr instanceof TypeLiteral) {
+	public static IAnnotationBinding obatinAnnotationBinding(BodyDeclaration declaration, Class<?> annotationClass) {
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.utils.eclipse/src/hu/elte/txtuml/utils/jdt/SharedUtils.java]: Typo, should be _obtain_.

File: dev/plugins/hu.elte.txtuml.utils.eclipse/src/hu/elte/txtuml/utils/jdt/SharedUtils.java [hunk 14]
```diff
@@ -102,11 +103,49 @@ public static Annotation obtainAnnotation(BodyDeclaration declaration, Class<?>
...
+				Annotation annotation = (Annotation) modifier;
+				if (identicalAnnotations(annotation, annotationClass)) {
+					return annotation.resolveAnnotationBinding();
+				}
+			}
+		}
+		return null;
+	}
+	
+	public static ITypeBinding obatinTypeBindingFromExpression(Expression expr) {
```
Comment (djnemeth) [inline dev/plugins/hu.elte.txtuml.utils.eclipse/src/hu/elte/txtuml/utils/jdt/SharedUtils.java]: Typo, should be _obtain_.

File: dev/tests/hu.elte.txtuml.export.cpp.tests/src/hu/elte/txtuml/export/cpp/wizardz/CompileTests.java [hunk 15]
```diff
@@ -183,7 +184,7 @@ public void exportTest() {
 		}
 	}
 
-	@Test(timeout=600000)
+	@Test(timeout=720 * 1000)
```
Comment (djnemeth) [inline dev/tests/hu.elte.txtuml.export.cpp.tests/src/hu/elte/txtuml/export/cpp/wizardz/CompileTests.java]: Is there some reasoning behind this limit?

File: examples/demo/pingpong/src/pingpong/j/cpp/PingPongConf.java [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+import hu.elte.txtuml.api.deployment.Group;
+import hu.elte.txtuml.api.deployment.Runtime;
+import hu.elte.txtuml.api.deployment.RuntimeType;
+import pingpong.j.model.Game;
+import pingpong.j.model.Player;
+import pingpong.j.model.Racket;
+
+@Group(contains = {Player.class, Game.class, Racket.class})
+@Runtime(RuntimeType.SINGLE)
+public class PingPongConf extends Configuration{
```
Comment (djnemeth) [inline examples/demo/pingpong/src/pingpong/j/cpp/PingPongConf.java]: Should be _PingPongConfiguration_. Please Ctrl + Shift + O and Ctrl + Shift + F as well.

File: examples/demo/pingpong/src/pingpong/j/cpp/main.cpp [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+#include "deployment.hpp"
+#include "runtime/Action.hpp"
+#include "EventStructures.hpp"
+#include "Game.hpp"
+
+int main()
+{
+  Env::initEnvironment();
+  UsedRuntimePtr rt = UsedRuntimeType::getRuntimeInstance();
+
```
Comment (djnemeth) [inline examples/demo/pingpong/src/pingpong/j/cpp/main.cpp]: Unnecessary newline. This is a demo, we have to make it pretty.

File: examples/demo/pingpong/src/pingpong/j/model/Player.java [hunk 18]
```diff
@@ -24,7 +23,6 @@
 	public class BallAtPlayerPort extends Port<BallIfc, BallIfc> {
 	}
 
-	@BehaviorPort
```
Comment (djnemeth) [inline examples/demo/pingpong/src/pingpong/j/model/Player.java]: Why was this necessary?

File: examples/demo/pingpong/src/pingpong/j/model/Racket.java [hunk 19]
```diff
@@ -66,7 +66,7 @@ public void effect() {
 
 		@Override
 		public boolean guard() {
-			return Else();
+			return getTrigger(Ball.class).countdown <= 0;
```
Comment (djnemeth) [inline examples/demo/pingpong/src/pingpong/j/model/Racket.java]: I have seen in the message of 4d8b652 that `else` cannot be exported. Does this mean that it hasn't been implemented yet, or do we face a theoretical problem? Either way, I suggest to add a comment here to remark that the original guard was `Else()`.

File: examples/demo/pingpong/src/pingpong/j/model/package-info.java [hunk 20]
```diff
@@ -1,4 +1,4 @@
-@Model
+@Model("PingPong")
```
Comment (djnemeth) [inline examples/demo/pingpong/src/pingpong/j/model/package-info.java]: If supported, please update _XPingPong_ as well to keep it consistent with _JPingPong_. What about _Clock_?

File: examples/demo/pingpong/src/pingpong/j/model/Player.java (hunk 18 above)
Comment (nagyandras95) [reply examples/demo/pingpong/src/pingpong/j/model/Player.java]: Because it is not a behavior port (out port cannot be behavior port anyway) but it is a delegation port to an lower level component. In case of delegation connection the parent port should not be behavior port so the generated cpp code does not complies in this case.

File: examples/demo/pingpong/src/pingpong/j/model/package-info.java (hunk 20 above)
Comment (nagyandras95) [reply examples/demo/pingpong/src/pingpong/j/model/package-info.java]: Colock contains more not supported elements so I would postpone to support it.

File: examples/demo/pingpong/src/pingpong/j/cpp/main.cpp (hunk 17 above)
Comment (nagyandras95) [reply examples/demo/pingpong/src/pingpong/j/cpp/main.cpp]: I have deleted the next empty line..

File: dev/tests/hu.elte.txtuml.export.cpp.tests/src/hu/elte/txtuml/export/cpp/wizardz/CompileTests.java (hunk 15 above)
Comment (nagyandras95) [reply dev/tests/hu.elte.txtuml.export.cpp.tests/src/hu/elte/txtuml/export/cpp/wizardz/CompileTests.java]: I have introduced the limit in order to avoid the test deadlock cased by cpp runtime dedlock.  I have added the pingpong model to tests so the timout was incarsed as well. It is only an upper estimate to the running time of cpp tests.

**Comment by djnemeth** [comment]
My previous remarks about the code were (mostly) addressed. I trust @kmozsi about the functionality. Let's merge! :shipit:
